### ISSUE TITLE
RaspyRFM HW support

### ISSUE THREAD

**Issue by Phunkafizer** [pilight/pilight#344]
Hey,

i would like to implement this radio-hardware in pilight:

http://www.seegel-systeme.de/produkt/raspyrfm-ii/

Can you give me a starting point how to add a new hardware device in pilight?

Stefan

**Comment by CurlyMoo** [comment]
I would just start with a working C program that does what you want to achieve. Then we can look at porting the whole to pilight.

**Comment by Phunkafizer** [comment]
Thank you for the reply. There are lots of implementations for the RFM69 radio module on the web. Meanwhile I investigated a bit more the other hardwarebindings of pilight, especially 433gpio.c. Very good to read by the way!
The main problem is: You give already "processed" pulse data to the hardware layer, e. g. 100, 400, 100, 400, 400, 100, 100, 400. Is is a good working way for all radio modules without a buildin bitprocessor, which need realtime data on a port pin. More advanced radiomodules (rfm69, cc1020, lots more) require a bitstream, which means for a 100, 400 you would send binary 0b10000 for instance. Of course this could be converted in this way, but can cause some strange behavior. It would be more safe if the data to the hardware layer is not passed as absolute timing values, but as multiplier for a timebase, e.g. (pseudocode):

hardware_send(int timebase, int data[]) {
processing for different hardware;
for pulsedevices (e.g. TX433) multiply every data value with timebase
for fifodevices, generate number of 1 und 0 bits according to data
}

int data[] = {1, 3, 1, 3, 3, 1, 3, 1, 3, 1, 3, 1, 1, 3, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 1, 3, 1, 3, 1, 10}
hardware_send(90, data);

for pulsedevices this would result in:
90, 270, 90, 270, 270, 90, 270, 90, 270, 90, 270, 90, 90, 270, 90, 270, 270, 90, 90, 270, 270, 90, 90, 270, 90, 270, 90, 270, 90, 900
for fifodevices this would result in:
1000 1000 1110 1110 1110 1110 0111 0111 1110 0111 1110 0111 0111 0111 10000000000

do you see a change that the layer to the hardwarelibs can be changed in this way?
This would also introduce the possibility to support way more hardware, e. g. also the brematic, itgw433, connair

Stefan

**Comment by CurlyMoo** [comment]
Have you checked the 433nano code which used a different approach.

And, ofc, you can add a pre-processor in the hardware module itself.

**Comment by Phunkafizer** [comment]
Yes I checked the 433nano code. It is indexing the different pulse lengths. This is a bit like putting salt in water and boiling it off again ;-). But yes, would be a way, nevertheless having the data for the hardwarelayer  consisting of the pulse length and an array of multipliers  would be much easier to handle

**Comment by CurlyMoo** [comment]
Can you create a proof-of-concept based on one of the existing protocols? For example archtech_switch?

**Comment by Phunkafizer** [comment]
Refering to this page:

https://wiki.pilight.org/doku.php/arctech_switch

But your raw-array would contain:

1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 1 1 5
1 5 1 1
1 1 1 5
1 5 1 1
1 1 1 5
1 1 1 5
1 1 1 5
1 5 1 1

and you would additionally pass a multiplier (timebase) to the hardwarelayer, in this case 275

In the 433gpio module you would simply multiply every member (unsigned char is sufficient by the way, would save memory) with the timebase to get your sending data. If you have a streaming-device like the raspyRfm, you would just send the amount of bits according to the raw data. And data for a intertechno lan gateway just has to be converted to a text line

**Comment by CurlyMoo** [comment]
And to the fifodevice example?

**Comment by Phunkafizer** [comment]
For fifodevice (bit data):

1 0000000000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 00000 1 0
1 0 1 00000
1 00000 1 0
1 0 1 00000
1 0 1 00000
1 0 1 00000
1 00000 1 0
1 0000000000000000000000000000000000000000

so fifo data stream for the RFM69 would look like:
uint8_t fifodata[] = {0x80, 0x14, 0x14, 0x14, .........}


And here is like a UDP string for the brematic / itgw433 / connair would look:

TXP:0,0,10,5600,275,132,1,10, 1, 1, 1, 5, 1, 1, 1, 5, ...........};
275 is timebase, 132 is number of pulses, following the pulse data

**Comment by CurlyMoo** [comment]
Beat me up, because i can't see how the fifo relates to the first example?

**Comment by Phunkafizer** [comment]
No worries ;-) Basically you are sending on/off pulses on the radio side (this is called OOK modulation). Let's take this pulsedata:

286 201 289 1337 

This means: 286 µS on, 201 µS off, 289 µS on, 1337 µS off
If you round it by the "pulsefrequency": 1 * 275 µS on, 1 * 275 µS off, 1 * 275 µS on, 5 * 275 µS off

When operating with FIFO radio modules, they send with configurable bitrate. In this case you would set it to 3636 bit/s (= 1/275 µS), so every 275 µS it would send out exactely one bit. If you put a 1 in the fifo, it will switch on the RF output for 275 µS. So if you want to "translate" the example above:

1 * 275 µS on -> 1
1 * 275 µS off -> 0
1 * 275 µS on -> 1
5 * 275 µS off -> 00000

**Comment by CurlyMoo** [comment]
I would still opt for a hardware module that translates the current pulsestreams to fifo streams just like i did with the usb nano. Another idea would be to adapt the USB nano code to a fifo input.

**Comment by Phunkafizer** [comment]
If the USB nano works, there is not really a reason to change it. 
OK, although the current dataformat is a bit "odd", I could calculate the greatest common divisor of the pulses and convert it to fifoformat. Another thing: when creating the sync pulse in createFooter, you use the macro PULSE_DIV. What is the reason for that name?

**Comment by CurlyMoo** [comment]
When the forum is back online i would suggest searching there. We've got that question quite a few times and explained it just so many times :smile:

**Comment by Phunkafizer** [comment]
ok, meanwhile some steps further, compiled the developer branch on my raspberry. What do I have to touch if have a new c/h file in the hardware folder? By the way, nice thing wiringX! Everything in there I need for the new module. Does the SPI work in wiringX?

**Comment by CurlyMoo** [comment]
Can we discuss writing a new hardware module on the forum so other can use that same information as well?
